java/flight/src/test/java/org/apache/arrow/flight/TestDictionaryUtils.java

```diff
@@ -0,0 +1,91 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.flight;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.TreeSet;
+
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.RootAllocator;
+import org.apache.arrow.vector.TinyIntVector;
+import org.apache.arrow.vector.dictionary.Dictionary;
+import org.apache.arrow.vector.dictionary.DictionaryProvider;
+import org.apache.arrow.vector.types.pojo.ArrowType;
+import org.apache.arrow.vector.types.pojo.DictionaryEncoding;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.types.pojo.FieldType;
+import org.apache.arrow.vector.types.pojo.Schema;
+import org.junit.Test;
+
+import com.google.common.collect.ImmutableList;
+
+/**
+ * Test cases for {@link DictionaryUtils}.
+ */
+public class TestDictionaryUtils {
+
+  @Test
+  public void testReuseSchema() {
+    FieldType varcharType = new FieldType(true, new ArrowType.Utf8(), null);
+    FieldType intType = new FieldType(true, new ArrowType.Int(32, true), null);
+
+    ImmutableList<Field> build = ImmutableList.<Field>builder()
+            .add(new Field("stringCol", varcharType, null))
+            .add(new Field("intCol", intType, null)).build();
+
+    Schema schema = new Schema(build);
+    Schema newSchema = DictionaryUtils.generateSchema(schema, null, new TreeSet<>());
+
+    // assert that no new schema is created.
+    assertTrue(schema == newSchema);
+  }
+
+  @Test
+  public void testCreateSchema() {
+    try (BufferAllocator allocator = new RootAllocator(1024)) {
+      DictionaryEncoding dictionaryEncoding =
+              new DictionaryEncoding(0, true, new ArrowType.Int(8, true));
+      TinyIntVector dictVec = new TinyIntVector("dict vector", allocator);
+      Dictionary dictionary = new Dictionary(dictVec, dictionaryEncoding);
+      DictionaryProvider dictProvider = new DictionaryProvider.MapDictionaryProvider(dictionary);
+      TreeSet<Long> dictionaryUsed = new TreeSet<>();
+
+      FieldType varcharType = new FieldType(true, new ArrowType.Utf8(), dictionaryEncoding);
+      FieldType intType = new FieldType(true, new ArrowType.Int(32, true), null);
+
+      ImmutableList<Field> build = ImmutableList.<Field>builder()
+              .add(new Field("stringCol", varcharType, null))
+              .add(new Field("intCol", intType, null)).build();
+
+      Schema schema = new Schema(build);
+      Schema newSchema = DictionaryUtils.generateSchema(schema, dictProvider, dictionaryUsed);
+
+      // assert that a new schema is created.
+      assertTrue(schema != newSchema);
+
+      // assert the column is converted as expected
+      ArrowType newColType = newSchema.getFields().get(0).getType();
+      assertEquals(new ArrowType.Int(8, true), newColType);
```

Comment:
I might be missreading this but this doesn't seem correct to me.  https://github.com/apache/arrow/blob/master/format/Schema.fbs#L299 indicates the type shouldn't change if it is dictionary encoded?

Reply:
I see, I think my misunderstanding comes from the setup of this test.  The dictVector above should be VarCharVector and the stringCol, should probably reflect the encoding type (at least that seems to be the difference between the toMemoryFormat type).

Reply:
You are right. VarCharVector is the raw type, and TinyInt is the encoded type.

Reply:
But that scenario would never happen in practice?

Reply:
Why not? If we have only a few distinct strings, we can encode them with a TinyIntVector.

Maybe the name "dictVec" is misleading? I will change it to "encodedVec".

Reply:
@emkornfield The ut is fixed. Please take a look. Thank you.